### DISCUSSION TITLE
Implementation of operation (TimeOnly - TimeOnly)

### DISCUSSION THREAD

**Discussion by dmitry-lipetsk** [dotnet/runtime#60056] [General] [answered]
Hello,

https://github.com/dotnet/runtime/blob/d2cef2fe4c91f23ae9058ca91ed55c0e65c06d0c/src/libraries/System.Private.CoreLib/src/System/TimeOnly.cs#L249-L255

Could anybody explain (or point the exists discussion) the reason of this implementation?

1. If you want return a **TimeSpan** (interval between time), then result may be negative. Because if (A < B), then (A - B)<0

2. If you want return a TIME (as you are say in comment _"Gives the elapsed **time** between two points"_), then result must be TimeOnly.

Current implementation returns **time** in data type for **interval**.

It is correct?

----
Reason of this question.

In provider for EFCore need map this operation to SQL.

DBMS (PG, Firebird) for expression
```sql
time '01:00:00' - time '03:00:00'
```
returns **negative** value: -02:00:00.

Subtract operator of TimeOnly returns positive value "22:00:00".

----
Thanks.

**Comment by Clockwork-Muse** [comment, marked as answer]
> If you want return a TimeSpan (interval between time), then result may be negative. Because if (A < B), then (A - B)<0

In the ISO world (and in reality), there is no such thing as a negative duration.  (Although I personally agree with you here, that it does make some operations easier).  This was debated at design time, and changing it now would be a massive breaking change.

> If you want return a TIME (as you are say in comment "Gives the elapsed time between two points"), then result must be TimeOnly.

"Time" has multiple uses in English - the phrase "time me", for instance, is asking for a duration, not the wall-clock time.  Hence why it's "elapsed time" in the method documentation.

> Reason of this question.

If the values are subtracted on the database side, the value returned will be a negative timestamp.  If you're emulating this in the EFCore mapping, you need to add an extra check.

**Comment by bartonjs** [comment]
> If you want return a TIME (as you are say in comment "Gives the elapsed time between two points"),

"elapsed time" is the same as "a span of time", so TimeSpan is correct for that one.

`TimeOnly` is half of `DateTime` (the time half).   It represents values that can be seen on a 24h clock.  23:59:59 + one second is 00:00:00.  It's for "this meeting is at 3, every Friday".

**Comment by svick** [comment]
When you have a `from` time and a `to` time and the `to` time is smaller, then you can either interpret that as a negative range, or as a positive range crossing midnight. As I understand it, `TimeOnly` always chooses the latter option, which is why `-` is always going to return a positive value.

The result is a `TimeSpan`, because that is the correct type: `TimeOnly` represents a time of day (e.g. midnight), whereas `TimeSpan` represents a duration.
